src/JSON/ZeroCopy/Deserializer.dfy

```diff
@@ -276,49 +305,110 @@ module {:options "-functionSyntax:4"} JSON.ZeroCopy.Deserializer {
       ensures pr.Success? ==> pr.value.StrictlySplitFrom?(cs0, BracketedSpec)
     {
       var elem :- Element(elems.cs, json);
-      var sep := Core.TryStructural(elem.cs);
-      var s0 := sep.t.t.Peek();
-      if s0 == SEPARATOR as opt_byte then
-        assert AppendWithSuffix.requires(open.cs, json, elems, elem, sep) by {
-          assert {:focus} elems.cs.StrictlySplitFrom?(json.cs);
-          assert elems.SplitFrom?(open.cs, SuffixedElementsSpec);
-          assert elem.StrictlySplitFrom?(elems.cs, ElementSpec);
-          assert sep.StrictlySplitFrom?(elem.cs, c => Spec.Structural(c, SpecView));
-          assert forall e | e in elems.t :: e.suffix.NonEmpty?;
-          assert {:split_here} true;
-        }
-        var elems := AppendWithSuffix(open.cs, json, elems, elem, sep);
-        Elements(cs0, json, open, elems)
-      else if s0 == CLOSE as opt_byte then
-        assert AppendLast.requires(open.cs, json, elems, elem, sep) by {
-          assert sep.StrictlySplitFrom?(elem.cs, c => Spec.Structural(c, SpecView));
-          assert elems.cs.StrictlySplitFrom?(json.cs);
-          assert elems.SplitFrom?(open.cs, SuffixedElementsSpec);
-          assert elem.StrictlySplitFrom?(elems.cs, ElementSpec);
-        }
-        var elems' := AppendLast(open.cs, json, elems, elem, sep);
-        assert elems'.SplitFrom?(open.cs, SuffixedElementsSpec) by {
-          assert elems'.StrictlySplitFrom?(open.cs, SuffixedElementsSpec);
-        }
-        var bracketed := BracketedFromParts(cs0, open, elems', sep);
-        assert bracketed.StrictlySplitFrom?(cs0, BracketedSpec);
-        Success(bracketed)
+      if elem.cs.EOF? then
```

Comment:
Just for my understanding, was the key improvement re-ordering the cases or was this a deeper optimization of some kind?

Reply:
The case split is new I believe. As you can see from the Lemma `AboutTryStructural` I added, it is important to know that `!elemen.cs.EOF?`.

Reply:
I moved the Lemma `AboutTryStructural` to the actual `assertion by` clauses it is needed for.